### PR TITLE
WIP: automation script

### PR DESCRIPTION
**Pull request by ilanasegall**
only tested locally. Can test out with command as follows:
python australis_report_gen.py -s 20140101 --release --end 20141212 --beta --aurora
WIP: not complete.

### CODE REVIEW

File: telemetry-server/mapreduce/job.py [hunk 1]
```diff
@@ -51,37 +51,38 @@ class Job:
 
     def __init__(self, config):
         # Sanity check args.
-        if config.num_mappers <= 0:
+        print config
```
Comment (bwinton) [inline telemetry-server/mapreduce/job.py]: Do we really want to print the config, and if so, do we want to do something like
`print >> sys.stderr, config`
so that we can redirect stdout to a file?

File: telemetry-server/australis_report_gen.py [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+'''
+-default before/after australis?
+- generate filters
+- generate two run.sh
+- run uitour.py/output, uitour_peruser.py/output_peruser
+- run processoutput/processoutput_peruser
+- generate stats
+'''
+
+import argparse
```
Comment (bwinton) [inline telemetry-server/australis_report_gen.py:12]: I'm not suggesting we change this, but I recently read "argh, click, docopt, and plac are all great alternatives to argparse/optparse/gflags", so perhaps those are things we might want to look into in the future…

File: telemetry-server/australis_report_gen.py [hunk 3]
```diff
@@ -0,0 +1,131 @@
+#!/usr/bin/python
+
+'''
+-default before/after australis?
```
Comment (bwinton) [inline telemetry-server/australis_report_gen.py:4]: Default after.  I think before is something we'll need to run once, and then possibly never again…  ;)

File: telemetry-server/australis_report_gen.py [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+- generate filters
+- generate two run.sh
+- run uitour.py/output, uitour_peruser.py/output_peruser
+- run processoutput/processoutput_peruser
+- generate stats
+'''
+
+import argparse
+import json
```
Comment (bwinton) [inline telemetry-server/australis_report_gen.py:13]: Did we want json or simplejson here?  (I remember one of them was faster, but I forget which…)

File: telemetry-server/australis_report_gen.py [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+
+from mapreduce.job import Job
+
+FILTER_DIR = "./"
+
+
+#fwhat input do we want??? incomplete
+def generate_filters(args):
+    chans = {key: args.__dict__[key] for key in ("nightly", "aurora", "beta", "release")}
+    channels = filter(lambda x: args.__dict__[x] is True, chans)
```
Comment (bwinton) [inline telemetry-server/australis_report_gen.py:29]: Could this be simplified to `channels = [channel for channel in args if args[channel] == True and channel in ("nightly", "aurora", "beta", "release")]`?

File: telemetry-server/australis_report_gen.py [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+    },
+    {
+      "field_name": "appUpdateChannel",
+      "allowed_values": channels
+    },
+    {
+      "field_name": "appVersion",
+      "allowed_values": {
+        "min": "29.0",
+        "max": "29.999"
```
Comment (bwinton) [inline telemetry-server/australis_report_gen.py:50]: Hmmm…  I wonder if we want this to be an option, and add something like:
`"min": args[version] + ".0", "max": args[version] + ".999"`

File: telemetry-server/australis_report_gen.py [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+      "field_name": "appBuildID",
+      "allowed_values": {
+        "min": "20140123004002",
+        "max": "99999999999999"
+      }
+    },
+    {
+      "field_name": "submission_date",
+      "allowed_values": {
+        "min": "20140203",
```
Comment (bwinton) [inline telemetry-server/australis_report_gen.py:63]: I'm almost certain that _these_ should be values we calculate.  :)

File: telemetry-server/australis_report_gen.py [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+        "max": "20140429"
+      }
+    }
+    ]
+    }
+    filterfile = FILTER_DIR + "filter-auto.json"
```
Comment (bwinton) [inline telemetry-server/australis_report_gen.py:69]: Related to the comment above, I wonder whether we want to save previous filters with semi-unique names, to make it easier to re-run older jobs.  (Something like "filter-auto-week-43.json"?)

File: telemetry-server/australis_report_gen.py [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+    with open(filterfile, "w") as outfile:
+        json.dump(fltr, outfile)
+
+    return filterfile
+
+#many of these args can be exposed at the command line. no need for now.
```
Comment (bwinton) [inline telemetry-server/australis_report_gen.py:75]: `#TODO:`!  ;)
(Should we file a github issue or something so that we don't forget to do this later?

File: telemetry-server/australis_report_gen.py [hunk 10]
```diff
@@ -0,0 +1,131 @@
...
+      job.mapreduce()
+  except:
+      traceback.print_exc(file=sys.stderr)
+      exit_code = 2
+  duration = timer.delta_sec(start)
+  print "All done in %.2fs" % (duration)
+  return exit_code
+
+
+parser = argparse.ArgumentParser()
```
Comment (bwinton) [inline telemetry-server/australis_report_gen.py:117]: I like to put the main logic in a `main` function, using the template from Guido van Rossum, at the bottom of [this page](http://www.artima.com/weblogs/viewpost.jsp?thread=4829)

File: telemetry-server/australis_report_gen.py [hunk 11]
```diff
@@ -0,0 +1,131 @@
...
+parser.add_argument("-s", "--start", help="enter start date in the format YYYYMMDD, ex: 20140203", required=True)
+parser.add_argument("-e", "--end", help="enter end date in the format YYYYMMDD, ex: 20140203", required=True)
```
Comment (bwinton) [inline telemetry-server/australis_report_gen.py:119]: Do we want start/end, or just a single "week number" which would contain that entire week's data?

File: telemetry-server/mapreduce/job.py [hunk 12]
```diff
@@ -51,37 +51,38 @@ class Job:
...
+        self._bucket_name = config.get("bucket")
+        self._aws_key = config.get("aws_key")
+        self._aws_secret_key = config.get("aws_secret_key")
+        modulefd = open(config.get("job_script"))
         # let the job script import additional modules under its path
-        sys.path.append(os.path.dirname(config.job_script))
+        sys.path.append(os.path.dirname(config.get("job_script")))
         ## Lifted from FileDriver.py in jydoop.
-        self._job_module = imp.load_module("telemetry_job", modulefd, config.job_script, ('.py', 'U', 1))
+        self._job_module = imp.load_module("telemetry_job", modulefd, config.get("job_script"), ('.py', 'U', 1))
```
Comment (bwinton) [inline telemetry-server/mapreduce/job.py]: Do we want to try getting this in the upstream telemetry-server code?  If so, perhaps we should put it in its own commit…

File: telemetry-server/australis_report_gen.py (hunk 3 above)
Comment (ilanasegall) [inline telemetry-server/australis_report_gen.py:4]: noted.

File: telemetry-server/australis_report_gen.py (hunk 2 above)
Comment (ilanasegall) [inline telemetry-server/australis_report_gen.py:12]: optparse is being deprecated, so I'd stay away from that one. Ow, using this as it's used across the codebase, but happy to use something else in the future (as argparse is more verbose than I would like).

File: telemetry-server/australis_report_gen.py (hunk 4 above)
Comment (ilanasegall) [inline telemetry-server/australis_report_gen.py:13]: Simplejson is faster (ujson being the fastest). Here it's such a small file that hopefully we wouldn't see a difference, but could always use optimization. Changed.

File: telemetry-server/australis_report_gen.py (hunk 2 above)
Comment (bwinton) [inline telemetry-server/australis_report_gen.py:12]: Yeah, I was reading about argh during lunch just now.  It seems pretty cool, but certainly not worth the time to switch this over.  :)

File: telemetry-server/australis_report_gen.py (hunk 5 above)
Comment (ilanasegall) [inline telemetry-server/australis_report_gen.py:29]: args is a Namespace object, and those aren't iterable, so unfortunately the [x for x in...] construct doesn't work here. Agreed that this is gross, and another good reason to move away from argparse.

File: telemetry-server/australis_report_gen.py (hunk 6 above)
Comment (ilanasegall) [inline telemetry-server/australis_report_gen.py:50]: love it. implemented.

File: telemetry-server/australis_report_gen.py (hunk 5 above)
Comment (bwinton) [inline telemetry-server/australis_report_gen.py:29]: Rats!  Wait, is args.**dict** iterable?

File: telemetry-server/australis_report_gen.py (hunk 7 above)
Comment (ilanasegall) [inline telemetry-server/australis_report_gen.py:63]: Yes. Given the channel and versions, do we even need these fields, or are they redundant? We can add these fields as optional arguments for the very delicate analysis later, but they don't seem to be necessary right now with the other fields specified (correct me if i'm wrong!!!) Should we just set these from 0 to 999999 by default?

File: telemetry-server/australis_report_gen.py (hunk 8 above)
Comment (ilanasegall) [inline telemetry-server/australis_report_gen.py:69]: Love it. I also want to do this for output. Can we maybe let the user "tag" the run and label everything (output, filter) as follows?

File: telemetry-server/australis_report_gen.py (hunk 7 above)
Comment (bwinton) [inline telemetry-server/australis_report_gen.py:63]: Probably redundant.  And I think we could just set it to `['*']`, maybe?

File: telemetry-server/australis_report_gen.py (hunk 8 above)
Comment (bwinton) [inline telemetry-server/australis_report_gen.py:69]: Seems like a decent plan.  Auto-tag it with the week number if nothing is specified, maybe?

File: telemetry-server/mapreduce/job.py (hunk 12 above)
Comment (ilanasegall) [inline telemetry-server/mapreduce/job.py]: I struggled for a while with the cleanest way to do this. Added a bug and let's discuss.
Perhaps the best way is to make it its own "non command line invoked job" class with a different name.

File: telemetry-server/australis_report_gen.py (hunk 11 above)
Comment (ilanasegall) [inline telemetry-server/australis_report_gen.py:119]: good question.
